compressors/lz77.py

```diff
@@ -0,0 +1,705 @@
+| zlib.wasm                           |86408     |43471             |37448      |
+| jquery-2.1.4.min.js                 |84345     |34447             |29569      |
+| random.bin (random bytes)           |1000000   |1004970           |1000328    |
+| repeat_As.txt                       |1000000   |208               |1004       |
+| kennedy.xls                         |1029744   |330252            |204004     |
+| alice29.txt                         |152089    |63376             |54416      |
+
+"""
+
+import argparse
+from dataclasses import dataclass
+import os
+import tempfile
+from typing import List, Tuple
+from compressors.elias_delta_uint_coder import EliasDeltaUintDecoder, EliasDeltaUintEncoder
+from compressors.huffman_coder import HuffmanDecoder, HuffmanEncoder
+from core.data_block import DataBlock
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from core.data_stream import Uint8FileDataStream
+from core.encoded_stream import EncodedBlockReader, EncodedBlockWriter
+from core.prob_dist import ProbabilityDist
+from utils.bitarray_utils import BitArray, bitarray_to_uint, uint_to_bitarray
+from utils.test_utils import (
+    create_random_binary_file,
+    try_file_lossless_compression,
+    try_lossless_compression,
+)
+
+ENCODED_BLOCK_SIZE_HEADER_BITS = (
+    32  # number of bits used to put a header consisting of encoded block size
+)
+DEFAULT_MIN_MATCH_LEN = 6
+DEFAULT_MAX_NUM_MATCHES_CONSIDERED = 64
+
+
+@dataclass
+class LZ77Sequence:
+    """LZ77Sequence that determines a series of operations during decompression.
+    - First copy `literal_count` literal characters to output.
+    - Next copy `match_length` characters from `match_offset` back in output to the output.
+    """
+
+    literal_count: int = (0,)
+    match_length: int = (0,)
+    match_offset: int = (0,)
+
+
+class LZ77Encoder(DataEncoder):
+    def __init__(
+        self,
+        min_match_length: int = DEFAULT_MIN_MATCH_LEN,
+        max_num_matches_considered: int = DEFAULT_MAX_NUM_MATCHES_CONSIDERED,
+        window_initialization: List = None,
+    ):
+        """LZ77Encoder. See module documentation for details.
+
+        Args:
+            min_match_length (int, optional): Minimum match length. Defaults to 6.
+            max_num_matches_considered (int, optional): Max number of matches considered
+                at a position to bound time complexity. The matches are considered
+                in order of recency and the longest match among the max_num_matches_considered
+                is chosen. Defaults to 64.
+            window_initialization (List, optional): initialize window (this is like side information
+                or dictionary in zstd parlance). The same initialization should be used for the decoder.
+
+        """
+        self.min_match_length = min_match_length
+        self.max_num_matches_considered = max_num_matches_considered
+        self.window = []
+        self.dict = (
```

Comment:
maybe name this variable a bit better. `self.dict` is not very descriptive

Reply:
substring_dict?

Reply:
done now